### DISCUSSION TITLE
some irbempy calls repeat for no obvious reason

### DISCUSSION THREAD

**Discussion by jvrod23** [spacepy/spacepy#479] [Support and Q&A] [answered]
I have a Windows 10 64-bit system.  I have installed Anaconda and am using the PyCharm IDE.  Python version is currently 3.7.3. 

I successfully installed SpacePy 0.2.1 a few months ago using pip install --upgrade spacepy --trusted-host pypi.org -- trusted-host files.pythonhosted.org.  I am not the administrator of this machine.

To become familiar with irbempy, I wrote a couple of toy scripts to try out some functions of greatest interest to me:  get_Lm, get_Lstar, and find_footpoint.  I encountered a strange behavior when running these scripts.  When I called get_Lm, or get_Lstar, the entire script would repeat from the top 6-10 times.  (There was no for-loop in the scripts.) The values output by these functions were identical every time.  This did not happen with find_footpoint.

A couple of days ago I updated SpacePy to 0.2.2.  Interestingly, this seemed to eliminate the problem with the toy scripts.  But, when I wrote a 'real' script to churn through POES data that calls 'get_Lstar', the problem resurfaced.   'get_Lstar' is called within a for-loop.  The first time through the loop, the entire script is called from the top, ~10 times. The second time and thereafter, just the calls within the loop are repeated, again ~10 times.  

Again, this does not happen when I replace 'get_Lstar' with 'find_footpoint'.

I updated numpy this morning from 1.16.2 to 1.19.3 but this had no obvious impact.

It's probably not caused by the use of the Ticktock or Coords classes since I use those for both 'get_Lstar' and 'find_footpoint'.

It doesn't seem to make a difference whether I run the script in debug or run mode.

It's a strange problem - I've never seen anything like this before.  I look forward to any insights you may have.

**Comment by jtniehof** [comment, marked as answer]
The Lstar functions use [multiprocessing](https://docs.python.org/3/library/multiprocessing.html) to parallelize the calculation. That means anything used by the interpreter needs to be re-imported in each process that is spawned. If there's any unconditional execution in there, they'll probably be executed again. [The docs describe this](https://docs.python.org/3/library/multiprocessing.html#the-spawn-and-forkserver-start-methods) but somewhat hide the fact that `spawn` is the only available method on Windows. The bottom line is I suspect you'll be fine if you protect the main code of the script like:

```python
if __name__ == '__main__':
    # do stuff
```

That also means any warnings that get raised by imports (anything imported at all) are also repeated. In particular #473 means importing `spacepy.time` will raise leapsecond warnings on 0.2.2.

The multiprocessing code is skipped entirely if `ncpus` is 1. You can set that in the [config file](https://spacepy.github.io/configuration.html). It'll mean lstar won't run as fast if doing multiple time steps, since it's on one CPU.

**Comment by jtniehof** [comment]
Ah, the relevant docs for Python 3 are [here](https://docs.python.org/3/library/multiprocessing.html#the-spawn-and-forkserver-start-methods)...it somewhat hides the issue, which is that `spawn` is the only method available on Windows.

**Reply by jvrod23** [reply]
Thanks.  I am not familiar with the use of multiprocessing.  I will look into this and let you know what I figure out.

**Reply by jtniehof** [reply]
I'm doing a terrible job of keeing this threaded...

I did a quick check. The multiprocessing code is skipped entirely if `ncpus` is 1. You can set that in [the config file.](https://spacepy.github.io/configuration.html). It'll mean lstar won't run as fast if doing multiple times, since it's on one CPU.

**Reply by jvrod23** [reply]
Preceding the main code with 'if __name__ == '__main__':' while keeping ncpus = 8 seemed to fix the problem, but the 'leap second' error message was generated 8 times when 'get_Lstar' was called.

When I successfully set ncpus = 1, this 'leap second' error message did not appear at all when 'get_Lstar' was called.  This was with and without 'if __name__ == '__main__':' preceding the main code. 

 I did find that I had to force a new 'spacepy.rc' to be generated (by deleting the old one) before I edited it in order for my edit to be recognized.  The existing 'spacepy.rc' had both SpacePy 0.2.2 and SpacePy 0.2.1 comments in it, and one section was repeated.

Anyways, thank you!  I think this problem has been solved.

**Reply by jtniehof** [reply]
The leapsecond warning is #473; you can ignore it. I updated my previous answer to capture all this so it's ready for reference.

Odd about the .rc file; the duplication of commented-out entries should be strictly cosmetic. It'll have some 0.2.1 and some 0.2.2 depending on when an option was introduced. Basically there's a comment with the default and the version it was from, then a commented line for easy editing.
